### ISSUE TITLE
💖 Forking to survive!

### ISSUE THREAD

**Issue by ssbarnea** [pycontribs/ruyaml#1]
I used ruamel.yaml in several project but the bitbucket repository removal and move to sourceforge seems to me a very weird change, especially with it history (not to mention the headaches it gives when you load it).

I suspect the reasons behind the move was the authors refusal to accept that "git" won the scm battle, many years ago. 

So I decided to fork it to gihub, where is much easier for others to contribute to. I think that 99% of python libraries I ever used are here, or at least have mirrors. 

I do not have any plans yet about the fork future, but I hope ruamel will finally decide to go github and benefit from its networking effect, so I could close the fork.

**Comment by EricCousineau-TRI** [comment]
Sounds like a well-intentioned, but perhaps too indirect, approach? (at least from looking at this post)

Did you already open an issue on the SourceForge page? If so, could you link it here (to show people you've done some level of due diligence?)
Perhaps a comment here would help? https://sourceforge.net/p/ruamel-yaml/tickets/314/

Some food for thought / peanut gallery comments:

- I did benefit from this, so thank you!
- At least one other person is doing this, perhaps join efforts? https://github.com/commx/ruamel-yaml (\cc @commx)
- What's your expected update frequency? @commx's repo seems more recently updated
- Could you add a link to the original SourceForce page anywho? https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/
- Could you update the Project Description to link to this issue?
- Not having an update frequency nor a concrete plan for the fork is rather confusing... Perhaps make this read only for now, remove issues, and move this to a brief Wiki page, or a Gist (where people can comment)?

**Comment by smurfix** [comment]
The issue at https://sourceforge.net/p/ruamel-yaml/tickets/314/ is no longer readable, at least for me, after I left this comment there:

> 
> On 15.07.20 13:32, Sorin Sbarnea wrote:
> > That is why I created https://github.com/pycontribs/ruamel-yaml back in February but sadly I did not had time to work on it. It would be much better if others would join the efforth.
> 
> Is that repository up to date?
> 
> Do we create yet another fork ("ruyaml"?), or is there some chance of
> getting upload rights to pypi?
> 
> What would be the next steps?
> 
> Frankly, my first act would be to reorganize the code ("mkdir src
> src/ruyaml ; git mv …") and replace the whole of setup.py so that people
> who might be motivated to improve things see something familiar –
> instead of wondering why there's a static-expression evaluator in there.
> 
> Instead of github I'd be mostly-OK with moving to Gitlab instead.

Today I wrote an email to ruamel.yaml's main contributor:

> I have to say that blocking discussion of alternate solutions instead
> of, say, talking about why you prefer to keep things as they are, or
> (possibly) why you think my comment was out of line, comes across as
> quite heavy-handed and unfriendly, esp. since the links in your README,
> and the RTD documentation, are still woefully out of date.
> 
> Anthon, I can thus only assume that you don't want anybody else in your
> playground. Oh well. Time for another fork, one of these days.

So yes I think it's time for a "real" fork.

**Comment by smurfix** [comment]
OK, I created a clone at https://github.com/smurfix/ruyaml and started to clean things up a bit (reorganize the directory structure, create a somewhat-standard `setup.py`, remove Py2 support, …).

Doesn't pass all of its tests yet but it's a start. I can either create a work-in-progress-ish Pull Request to here, or somebody might give me write access. ;-)

**Comment by ssbarnea** [comment]
@smurfix Super! I sent you an invite. Maybe the best approach would be to integrate your changes and to ping me to enable CI/CD jobs. I would suggest to do one change at a time and use the review process, so nobody would argue that we make deliberate changes.

Anyone interested in helping is welcomed to join.

**Comment by smurfix** [comment]
Hmm, if you set up CI/CD correctly, every pull request gets that done automatically. "One change at a time" is difficult because using a reasonable setup.py depends on moving the files and CI then depends on getting the tests to pass, which they don't currently.

If you tell me who you are on pypi I can add you to the "ruyaml" package I created (just to block the name, for now …)

I'll create a PR with what I have done so far.

**Comment by ssbarnea** [comment]
@smurfix Can you join #pycontribs on irc to sync-up a little bit on it? I see you are already made a change and conflicts with mine. 

Indeed keeping the source in root of repo is probably one of the worst decisions I ever seen for a python library, that is a PITA to fix but we need to do it.

**Comment by ssbarnea** [comment]
@smurfix I also created https://readthedocs.org/projects/ruyaml/ -- what is your rtd user to add you as a backup?

**Comment by smurfix** [comment]
I'm @smurf on RTD. 

Done.

**Comment by smurfix** [comment]
@AvdN did not respond to my email about locking the Sourceforge bug, assuming that he got it in the first place.

At this point I'd be somewhat surprised if he'd consent to any of what we're doing, but I have been surprised before, so …

I'll prepare pull requests for parts of my PR. We'll see how git deals with all the renaming. ;-)

**Comment by jwilson8767** [comment]
Awesome! Definitely need this! Does anyone have an idea how we can preserve / some way archive the old bitbucket issues? I'm worried they're gone gone already!

**Comment by layday** [comment]
The issues were transferred to SourceForge.

**Comment by ssbarnea** [comment]
I did not had time to work more on this but seeing others interested can easily boost my interest, especially as I was displeased about the censorship and lack of reply on emails. Just say you want to help you will receive an invite to this repo.

**Comment by jwilson8767** [comment]
@ssbarnea I most definitely want to help!

**Comment by jwilson8767** [comment]
> The issues were transferred to SourceForge.

Perfect thanks! I have made a copy of them as of today (just using the SourceForge REST API), so we can preserve them later / raid them for things that should be in the docs.

**Comment by jwilson8767** [comment]
Alright, so I've spent a bit more time thinking about the future of this fork. I'd like to outline a few goals for comment and the steps that I think will be needed to accomplish them. However, before I get to that, I want to point out a few of **ruamel.yaml's  strengths today**: 
1. Supports parsing, mutation, and dumping of almost every feature of the YAML 1.2 spec.
2. Most notably, when using the round-trip loader, comments are preserved at every stage
3. Even though virtually every value has to be wrapped in a special class/type, it's still relatively easy to use them as you would in native python if reading and basic mutation is all you need to do.
4. It has had years to mature and I don't hesitate to use it in production codebases.
5. It supports Python 3 and also Python 2 (which can also be a weakness).

With such an amazing foundation, **there's a lot of room for ruyaml to succeed as a community-lead project**. Anthon as really done the "zero-to-one" for us here, but it will come down to us to make this project the best it can be.

### Proposed goals for ruyaml:
#### Easy migration process
Whether a project is coming from plain JSON, PyYAML, or ruamel.yaml, it should be super easy to start using ruyaml. This may take a few "find & replace" operations, but shouldn't take more than an hour in most cases and should be well documented in the form of user guides. 

#### Complete documentation
One of the biggest strengths of YAML is that it's *human*-friendly. In order for ruyaml to be similarly friendly, it must be well documented. Beyond user guides, configuration and extensibility must be documented. This was a major complaint of ruamel.yaml and I hope we can do better.

#### Slow and stable release cycle
Let's face it, yaml parsing/dumping isn't what really excites me. What *does* excite me is dynamic workflows that intermingle some data / config, my notes / comments, and (python) code acting upon and mutating the data. With ruyaml, this can all happen in harmony! However, in order for that harmony to last, I suggest we limit breaking releases to only a couple of times per year.

### Strategic development plan
In order to meet the goals set out above, I suggest we take the following actions (approximately in order):
1. Drop Python 2 support ASAP! It adds additional complexity and moving forward is not necessary. I know some parts of this are done already, but there are still lots of remnants. Unless anyone has a differing opinion, I think we should target Python 3.6+.
2. Make a hard-break from the "ruamel.yaml" name and version numbering. The project is "ruyaml" and the first version should be `1.0.0` This will make documentation writing much easier. For posterity, I recommend stashing the current version of ruamel.yaml on just a `legacy` branch. 
3. Start writing in-code documentation to explain what all the parts of the project are and what they do. This then leads into the next step.
4. Revamp documentation with a focus on usability and on featuring all the little nuggets of customizability already in the codebase. I could imagine having a complete API map would be a start, but due to the number of "oh, just override this one thing and it'll do what you want" that is baked into the code base, I think there will be some creativity needed for this.
5. Release  version `1.0.0` to (hopefully) start being adopted and getting stronger community engagement.


Still lots to figure out, but hopefully we can iterate on the goals / planning here to make a better yaml experience.

**Comment by smurfix** [comment]
Step 2 is the most intrusive, all that renaming wreak havoc with merging other changes, thus it should be done first.

I'll work on that today.

**Comment by smurfix** [comment]
What/who should the package's author (and their email) be?

**Comment by jwilson8767** [comment]
I suggest we add a several of us, the maintainers both in the setup.py "author" field, and at the top of the license file (but leave Anthon listed as the original author). However, I think this is really a question for @ssbarnea.

**Comment by smurfix** [comment]
Anthon isn't the original author either; that distinction belongs to whoever wrote pyyaml …

@all: any objections to merging my initial rename PR?

**Comment by ssbarnea** [comment]
I would replace both author/maintainer with "PyContribs (Team)" and use the mailing list as contact email, there is no singular author/maintainer.

I have no problem listing Anthon as author in metadata but I doubt he wants it, clearly we want him documented, we do not plan to hide the fact this this started as a fork.

Ping me as often as needed and I will do my best to unblock you. The best part about this "move" is that the future should no longer depend on single person.

Regarding rename: i was not very keen about the rename but I released we cannot really progress without doing it, mainly due to the very weird way the library was developed in the first place.

**Comment by smurfix** [comment]
… ? mailing list?

**Comment by ssbarnea** [comment]
https://groups.google.com/g/pycontribs -- covers all projects under PyContribs and is not under high usage.

**Comment by jwilson8767** [comment]
> I would replace both author/maintainer with "PyContribs (Team)" and use the mailing list as contact email, there is no singular author/maintainer.
👍 

> Ping me as often as needed
What's the best way to contact you? Gitter?

> Regarding rename: i was not very keen about the rename but I realised we cannot really progress without doing it, mainly due to the very weird way the library was developed in the first place.

Exactly my thoughts.

**Comment by ssbarnea** [comment]
`#pycontribs` channel is on freenode, like most opensource projects. I am sending >2h/day with irc window focused (not proud of it, but part of my job). My nick there is `@zbr`.

**Comment by fopinappb** [comment]
I started using the original package recently and, while struggling to understand why anyone would be on SF these days, bumped into this.
One comment about the description: `elusive` is quite strong IMO and made me google for anything bad that had happened to people using the original package 🤦 
Finally decided to look through issues to understand if there was any explanation regarding the adjective and this one does make it clear indeed!

So I'd probably recommend adding reference to this thread from the README directly in some `What the fork?` section or alike (seems README needs some attention actually).

On another note, is there a plan to keep tabs on the original package changes to cherrypick/retrofit the changes into this one later on?
Maybe a reference branch with exact snapshot in sourceforce plus a schedule github action to pull from there and open pull request (on that reference branch) with any changes?
That'd make it easier to review, choose changes to apply in this package (in a different PR to master) and then merge that separate PR into the reference branch (for cleaner future changesets)

**Comment by fopina** [comment]
I started using the original package recently and, while struggling to understand why anyone would be on SF these days, bumped into this.
One comment about the description: elusive is quite strong IMO and made me google for anything bad that had happened to people using the original package facepalm
Finally decided to look through issues to understand if there was any explanation regarding the adjective and this one does make it clear indeed!

So I'd probably recommend adding reference to this thread from the README directly in some What the fork? section or alike (seems README needs some attention actually).

On another note, is there a plan to keep tabs on the original package changes to cherrypick/retrofit the changes into this one later on?
Maybe a reference branch with exact snapshot in sourceforce plus a schedule github action to pull from there and open pull request (on that reference branch) with any changes?
That'd make it easier to review, choose changes to apply in this package (in a different PR to master) and then merge that separate PR into the reference branch (for cleaner future changesets)

**Comment by ssbarnea** [comment]
Cherry picking is not possible, project major code layout changes that makes impossible to backport changes this way. The only option is to manually redo the work.

That was not initial desire but if you look at how the original project codebase was arranged you will understand. Also, you will need a drink or two after that.

**Comment by smurfix** [comment]
Well, if we're lucky the cherrypicking will magically find the code's new location.

I second the drink option.

**Comment by fopina** [comment]
I'll definitely go through the codebase if it's socially acceptable to drink after!

I meant more on keeping tabs on bug fixing introduced there than actually commit cherry-picking sorry

**Comment by ssbarnea** [comment]
I would really appreciate if someone can fix the docs building. I will take care of the github integration and assure we have a working CI pipeline.

**Comment by richm** [comment]
@pcahyna @nhosoi @spetrosi this may be an issue in the future with our role2collection conversion process - or at least we will need to know where to get the latest up-to-date source

**Comment by Skylion007** [comment]
Do we have any way of pulling changes from the upstream SourceForge? Seems like there have been useful features added that we may want to contribute here.

**Comment by ssbarnea** [comment]
Not automated, but if someone can do them it would be great. Dropping the problematic namespace packaging made backporting quite hard, likely easier to cherry pick changes.

**Comment by smurfix** [comment]
I'll take a stab at merging the stuff.

**Comment by smurfix** [comment]
@ssbarnea We have a problem. The initial Mercurial-to-git import you did uses github-mangled email addresses in its commit messages, while running an `hg-export | git-import` from the Sourceforge repository results in unmangled addresses. This results in your and my clones having no common base, thus the repositories cannot be merged.

The most straightforward solution would be for you to do an incremental HG import / another HG import with the same options you used before, and push that into an "upstream" branch. (Maybe create a cron job that does this weekly?)

If that is not possible and/or doesn't result in an identical tree, I can use my import and rebase the changes since your import onto yours. The problem is that doing this guarantees that we can never automate that import in the future.

**Comment by fopina** [comment]
@smurfix when I mentioned [that](https://github.com/pycontribs/ruyaml/issues/1#issuecomment-726807882) in the past I understood you guys did not want to follow the original project due to poor code quality. What changed?

**Comment by Skylion007** [comment]
This project is early enough that maybe it's okay to branch the new tree as
long as we have an automated way of pulling in commits? So maybe we should
rebase the entire repo onto the new tree. It seems like it would be better
to have non-mangled emails anyway, right?

**Comment by smurfix** [comment]
@Skylion007 We could do that, it's just a force-push (or a separate branch) either way, the problem is how to clean up the history with the least effort. Right now I'm working around the issue with a `.git/info/grafts` file but that's obviously not a future-proof solution.

@fopina If the original project does cleanup (like removal of more Py2-isms), bugfixes, more tests, or new features, we do want them. It's not black and white.

**Comment by ssbarnea** [comment]
I am not really sure what was lost especially as almost all commits were done by a single individual in the original project. While in general I would try to avoid divergences, I reached a point where I think we should stop looking backwards. I really want to avoid any extra drama and I crealy do not have time to deal with that.

@fopina code quality was not the main issue and based on the fact that I seen numerous people getting into troubles with the project, I doubt the non-technical problems are going to ever be addressed.

If it was not about the namespace issue and the need to move all the code I would have supported keeping close to the original project. 

@smurfix What would be the real benefit of rewriting history? I am doing my best to avoid it unless there is a serious reason/benefit for doing it. We all know how dangerous such a move is, even if we have only 9 public forks.

**Comment by smurfix** [comment]
@ssbarnea The benefit would be that merging upstream, now or in the future, wouldn't require rebasing their updates onto our tree.

That being said, it's not that big a hassle and just requires care and/or reasonable tooling, so if you can't incrementally update the original import (how was that done anyway? Did you ask github to pull from Sourceforge's Mercurial repo?) that's probably the best way forward.

**Comment by ssbarnea** [comment]
We would not be be able to cherrypick code from upstream, due to code layout differences. I think I used github import options. I would have done it locally only if that did not work.

**Comment by smurfix** [comment]
@ssbarnea I already merged upstream and am in the process of resolving the merge conflicts (mainly due to renaming to `ruyaml`). Otherwise the process looks reasonable, git is pretty good at figuring out file renames these days.

Apparently github's import mangles email addresses. Bah.

**Comment by smurfix** [comment]
Should be OK, tests pass and all that. Somebody might want to look at this before I merge to master.

**Comment by Skylion007** [comment]
Looks like it didn't push to update the pypy? Was that job suppose to fail?

**Comment by ssbarnea** [comment]
A new release is made using release option in github not by pushing tags. I made `0.90.0.2`, lets hope it will succeed publishing.

**Comment by Skylion007** [comment]
Already some more bugfixes import bugfixes in upstream we should pull.

**Comment by smurfix** [comment]
Yeah. Except those which we shouldn't.
```
<    from typing import Any, Dict, Optional, List  # NOQA
>    from typing import Any, Dict, Optional, List, Optional  # NOQA
```
This is exactly why code reviews are a good idea.

I'll do a PR shortly.

**Comment by Skylion007** [comment]
@smurfix We may want to pull from upstream again.

**Comment by smurfix** [comment]
That was quick; I managed to teach git about the previous merges.
https://github.com/pycontribs/ruyaml/pull/59

**Comment by gdubicki** [comment]
> Make a hard-break from the "ruamel.yaml" name and version numbering. The project is "ruyaml" and the first version should be 1.0.0 This will make documentation writing much easier. For posterity, I recommend stashing the current version of ruamel.yaml on just a legacy branch.

What happened to this proposal? One of the problems I had with the upstream ruamel.yaml is the non-semver versioning scheme. I was hoping that this project will follow it.

**Comment by gdubicki** [comment]
I have tried to describe what I propose we do with the versioning here: https://github.com/pycontribs/ruyaml/discussions/64

**Comment by ssbarnea** [comment]
I want to start using codeowners feature to automatically add two reviewers on pull-requests. If you are willing to help doing reviews please add a thumb-up to this comment, so I can compute the list of people to list there.

**Comment by ssbarnea** [comment]
I will close this task as I consider that survival is no longer an issue but I will be more than happy to welcome new contributors, feel free to continue addressing your interest on this ticket if you want to join maintenance efforts.
